File: symmetric-tree/symmetric-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        

        if(root->left == NULL && root->right == NULL){
            return true;
        }else if(root->left == NULL || root->right == NULL){
            return false;
        }
        if(root->left->val != root->right->val){
            return false;
        }
        
        deque<TreeNode*> q1;
        deque<TreeNode*> q2;

        q1.push_front(root->left);
        q2.push_back(root->right);

        TreeNode* leftsub;
        TreeNode* rightsub;

        while(!q1.empty()){
            if(q1.size()!=q2.size()){
                return false;
            }
            int len = q1.size();
            while(len--){
                leftsub = q1.front();q1.pop_front();
                rightsub = q2.back();q2.pop_back();
                if(leftsub->left != NULL && rightsub->right != NULL){
                    if(leftsub->left->val != rightsub->right->val){
                        return false;
                    }else{
                        q1.push_front(leftsub->left);
                        q2.push_back(rightsub->right);
                    }
                }else if(leftsub->left == NULL && rightsub->right == NULL){
                   
                }else if(leftsub->left == NULL || rightsub->right == NULL){
                    return false;
                }

                if(leftsub->right != NULL && rightsub->left != NULL){
                    if(leftsub->right->val != rightsub->left->val){
                        return false;
                    }else{
                        q1.push_front(leftsub->right);
                        q2.push_back(rightsub->left);
                    }
                }else if(leftsub->right == NULL && rightsub->left == NULL){
                   
                }else if(leftsub->right == NULL || rightsub->left == NULL){
                    return false;
                }
                
            }
        }

        return true;

    }
};

File: 101-symmetric-tree/symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):

    def level_order(self, root1, root2):
        if(root1==None and root2==None):
            return True
        elif(root1 == None or root2 == None):
            return False
        
        if(root1.val != root2.val):
            return False

        res1 = self.level_order(root1.left, root2.right)
        res2 = self.level_order(root1.right,root2.left)

        return res1 and res2


    def isSymmetric(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: bool
        """
        #return self.level_order(root.left,root.right)
        
        
        q1 = deque()
        q2 = deque()
        q1.append(root.left)
        q2.append(root.right)

        while(q1 and q2):
            size1= len(q1)
            size2 = len(q2)
            if(size1 != size2):
                return False

            while(size1):
                root1 = q1.popleft()
                root2 = q2.popleft()
                if(root1 == None and root2 == None):
                    size1 = size1-1
                    continue
                elif(root1 == None or root2 == None):
                    return False
                
                if(root1.val != root2.val):
                    return False
                
                q1.append(root1.left)
                q1.append(root1.right)
                q2.append(root2.right)
                q2.append(root2.left)
                size1 = size1-1
        
        return True

                
